internal/test/framework/framework.go

```diff
@@ -294,6 +300,62 @@ func (f *TestFramework) GetNode(externalIP string) (*v1.Node, error) {
 	return matchedNode, nil
 }
 
+// GetNode uses external IP and finds out the name associated with the node
+func (f *TestFramework) GetNodeName(externalIP string) (string, error) {
+	node, err := f.GetNode(externalIP)
+	if err != nil {
+		return "", fmt.Errorf("error while getting required kubernetes node object: %v", err)
+	}
+	return node.Name, nil
+}
+
+// RetrieveArtifacts should retrieve artifacts related the test run. Ideally this should retrieve all the logs related
+// to the Windows VM. This shouldn't return an error but should print the failures as log collection is nice to have
+// rather than a must have.
+// TODO: Think about how we can retrieve stdout from ansible out within this function
+func (f *TestFramework) RetrieveArtifacts() {
+	for i, vm := range f.WinVMs {
+		if vm == nil {
+			continue
+		}
+		if vm.GetCredentials() == nil {
+			log.Printf("error while getting credentials of the %d vm", i)
```

Comment:
There is no err per se but that the credentials were empty. `no credentials provided for vm %d` would be more accurate. This applies to the other logs here.